### DISCUSSION TITLE
Source for Avalonia.BuildServices is missing

### DISCUSSION THREAD

**Discussion by corngood** [AvaloniaUI/Avalonia#16878] [General]
### Describe the bug

I'm trying to package Avalonia for nixpkgs, and I can't find the source for Avalonia.BuildServices anywhere.

Apologies if I'm missing something, but I've searched the github org, etc.

### To Reproduce

Search the internet for "Avalonia.BuildServices".

### Avalonia version

11.0.11

**Comment by maxkatz6** [comment]
This package is not open sourced.

**Reply by Maxwell175** [reply]
That is crazy. 

Looking at that [package's contents](https://nuget.info/packages/Avalonia.BuildServices/0.0.31), it seems it adds the following task:

```
    <Target Name="AvaloniaStats"
            BeforeTargets="CoreCompile"
            Condition="$(DesignTimeBuild) != 'true' And $(UsedAvaloniaProducts) != '' And $(TargetFramework) != ''">
        <AvaloniaStatsTask
                ProjectName="$(MSBuildProjectName)"
                TargetFramework="$(TargetFramework)"
                RuntimeIdentifier="$(RuntimeIdentifier)"
                AvaloniaPackageVersion="$(AvaloniaMainPackageVersion)"
                OutputType="$(OutputType)"
        />
    </Target>
```

So, this closed-source telemetry task runs on every build and can do anything it wants. There is no way to prevent this task from being executed based on the conditions. So, we have to simply "trust me bro" that it will always be disarmed internally by the environment variable, that the Project Name is actually hashed, and that future versions of this closed source package will not decide to do anything else later on.

Similar projects in the space do [open source their telemetry](https://github.com/unoplatform/uno/blob/0a56fe88d7b4179f0b6623cef9a8bf53e07d9d11/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.Telemetry.cs#L28), which allows them to be audited to allow users to be confident in exactly what is being reported and it can easily be disabled with an **MSBuild** property. The big difference between environment variables and using MSBuild properties is that MSBuild properties can be set in the csproj file once and for all as opposed to environment variables which require more platform-specific configurations, which might make it easy to miss on one of the developers's machines.

**Reply by Maxwell175** [reply]
To audit this package for myself, I decompiled the libraries in this package. They are available here: https://github.com/Maxwell175/Avalonia.BuildServices

It seems that at least as of version 0.0.31, this task shorts out if either `AVALONIA_TELEMETRY_OPTOUT` or `NCrunch` environment variables are present: https://github.com/Maxwell175/Avalonia.BuildServices/blob/master/Avalonia.BuildServices/Avalonia/Telemetry/AvaloniaStatsTask.cs#L37

**Reply by kebox7** [reply]
> It seems that at least as of version 0.0.31, this task shorts out if either `AVALONIA_TELEMETRY_OPTOUT` or `NCrunch` environment variables are present:

This behavior is described in the official documentation: [Telemetry](https://github.com/AvaloniaUI/Avalonia/wiki/Telemetry)

**Reply by Maxwell175** [reply]
> This behavior is described in the official documentation: [Telemetry](https://github.com/AvaloniaUI/Avalonia/wiki/Telemetry)

It is irrelevant what the documentation says. When it comes to a security concern, what matters is the ability to audit the code to see what it _really_ does.

**Comment by corngood** [comment]
So I guess this is a telemetry component that can be removed without any problem?

The metadata lists it as MIT licenced.  I realise this doesn't require you to distribute the source, but I think it's surprising that something listed as MIT on nuget.org isn't open-source.

**Reply by rabbitism** [reply]
I think MIT license describes the repository, not the dependencies of the repository.

**Reply by corngood** [reply]
I meant the metadata on nuget.org, which is unrelated to the repository: https://www.nuget.org/packages/Avalonia.BuildServices

**Comment by corngood** [comment]
If before building Avalonia, I run:

```
        dotnet remove packages/Avalonia/Avalonia.csproj package Avalonia.BuildServices
```

The dependency is removed from the output nupkg, and `Avalonia.BuildServices` is no longer required to build Avalonia itself.  Projects consuming Avalonia still appear to build and run properly without pulling in `Avalonia.Buildservices`.

It'll be difficult for users currently using nuget.org to take advantage of this.

Maybe we need an open source alternative to nuget.org (along the lines of fdroid), where we can be confident about the source.

**Comment by stevemonaco** [comment]
The package was recently open sourced, in case anyone stumbles upon this discussion instead of the newer one (https://github.com/AvaloniaUI/Avalonia/discussions/19466).

Package: https://github.com/AvaloniaUI/Avalonia.BuildServices
